chronos/achannels.nim

```diff
@@ -0,0 +1,254 @@
+#  Apache License, version 2.0, (LICENSE-APACHEv2)
+#              MIT license (LICENSE-MIT)
+import strutils
+import asyncloop, handles, transport, asyncsync
+
+const hasThreadSupport* = compileOption("threads")
+
+when hasThreadSupport:
+  import locks
+
+type
+  RawAsyncChannelImpl {.pure, final.} = object
+    rd, wr, count, mask: int
+    maxItems: int
+    refCount: int
+    data: ptr UncheckedArray[byte]
+    when hasThreadSupport:
+      lock: Lock
+    eventNotEmpty: AsyncThreadEvent
+    eventNotFull: AsyncThreadEvent
+
+  RawAsyncChannel = ptr RawAsyncChannelImpl
+  AsyncChannel*[Msg] = RawAsyncChannel
+  AsyncChannelError* = object of CatchableError
+
+proc initLocks(rchan: RawAsyncChannel) {.inline.} =
+  ## Initialize and create OS locks.
+  when hasThreadSupport:
+    initLock(rchan.lock)
+  else:
+    discard
+  rchan.eventNotEmpty = newAsyncThreadEvent()
+  if rchan.maxItems > 0:
+    rchan.eventNotFull = newAsyncThreadEvent()
+
+proc deinitLocks(rchan: RawAsyncChannel) {.inline.} =
+  ## Deinitialize and close OS locks.
+  when hasThreadSupport:
+    deinitLock(rchan.lock)
+  else:
+    discard
+
+  close(rchan.eventNotEmpty)
+  if rchan.maxItems > 0:
+    close(rchan.eventNotFull)
+
+proc acquireLock(rchan: RawAsyncChannel) {.inline.} =
+  ## Acquire lock in multi-threaded mode and do nothing in
+  ## single-threaded mode.
+  when hasThreadSupport:
+    acquire(rchan.lock)
+  else:
+    discard
+
+proc releaseLock(rchan: RawAsyncChannel) {.inline.} =
+  ## Release lock in multi-threaded mode and do nothing in
+  ## single-threaded mode.
+  when hasThreadSupport:
+    release(rchan.lock)
+  else:
+    discard
+
+proc newAsyncChannel*[Msg](maxItems: int = -1): AsyncChannel[Msg] =
+  ## Create new AsyncChannel[Msg] with internal queue size ``maxItems``.
+  ##
+  ## If ``maxItems <= 0`` (default value), then queue size is unlimited.
+  let loop = getGlobalDispatcher()
+  result = cast[AsyncChannel[Msg]](allocShared(sizeof(RawAsyncChannelImpl)))
+  result.mask = -1
+  if maxItems <= 0:
+    result.maxItems = -1
+  else:
+    result.maxItems = maxItems
+  result.count = 0
+  result.wr = 0
+  result.rd = 0
+  result.refCount = 0
+  result.data = cast[ptr UncheckedArray[byte]](0)
+  result.initLocks()
+
+proc raiseChannelClosed() {.inline.} =
+  var err = newException(AsyncChannelError, "Channel closed or not opened")
+  raise err
+
+proc raiseChannelFailed() {.inline.} =
+  var err = newException(AsyncChannelError, "Channel synchronization failed")
+  raise err
+
+proc open*[Msg](chan: AsyncChannel[Msg]) =
+  ## Open channel ``chan``.
+  let loop = getGlobalDispatcher()
+  chan.acquireLock()
+  inc(chan.refCount)
+  chan.releaseLock()
+
+proc close*[Msg](chan: AsyncChannel[Msg]) =
+  ## Close channel ``chan``.
+  chan.acquireLock()
+  if chan.refCount == 0:
+    if not(isNil(chan.data)):
+      deallocShared(cast[pointer](chan.data))
+    chan.deinitLocks()
+    deallocShared(cast[pointer](chan))
```

Comment:
What happens if another thread tries to acquire the lock between

```
chan.deinitLocks()
deallocShared
```

Will it deadlock or crash?

I feel like close should start with `if chan.isNil: return` in case the thread lost a race.

Reply:
If API is going to be used properly this part of code will be executed only when all the threads already performed `channel.close()`. In such way there no more threads which will going to use channel.